ex/auto/writer/odev_build_form/build_form.py

```diff
@@ -0,0 +1,447 @@
+from __future__ import annotations
+from pathlib import Path
+
+from com.sun.star.awt import ActionEvent
+from com.sun.star.awt import FocusEvent
+from com.sun.star.awt import ItemEvent
+from com.sun.star.awt import MouseEvent
+from com.sun.star.awt import TextEvent
+from com.sun.star.awt import XActionListener
+from com.sun.star.awt import XButton
+from com.sun.star.awt import XControlModel
+from com.sun.star.awt import XFocusListener
+from com.sun.star.awt import XItemListener
+from com.sun.star.awt import XListBox
+from com.sun.star.awt import XMouseListener
+from com.sun.star.awt import XTextComponent
+from com.sun.star.awt import XTextListener
+from com.sun.star.awt import XWindow
+from com.sun.star.beans import PropertyChangeEvent
+from com.sun.star.beans import XPropertyChangeListener
+from com.sun.star.beans import XPropertySet
+from com.sun.star.document import XEventListener
+from com.sun.star.form import XGridControl
+from com.sun.star.form import XGridControlListener
+from com.sun.star.lang import EventObject
+from com.sun.star.sdbc import XResultSet
+from com.sun.star.text import XTextDocument
+from com.sun.star.view import XSelectionChangeListener
+from com.sun.star.view import XSelectionSupplier
+from com.sun.star.form import XForm
+import unohelper
+
+from ooo.dyn.form.form_component_type import FormComponentType
+
+from ooodev.office.write import Write
+from ooodev.utils.forms import Forms
+from ooodev.utils.gui import GUI
+from ooodev.utils.info import Info
+from ooodev.utils.lo import Lo
+from ooodev.utils.props import Props
+from ooodev.utils.file_io import FileIO
+
+# from .....resources import __res_path__
+# endregion imports
+
+class BuildForm(
+    unohelper.Base,
+    XEventListener,
+    XPropertyChangeListener,
+    XActionListener,
+    XTextListener,
+    XFocusListener,
+    XItemListener,
+    XMouseListener,
+    XSelectionChangeListener,
+    XGridControlListener,
+):
+    doc: XTextDocument = None
+
+    def __init__(self, db_path: Path) -> None:
+        super().__init__()
+
+        self._db_fnm = db_path
+
+        loader = Lo.load_office(Lo.ConnectSocket())
+        BuildForm.doc = Write.create_doc(loader)
+
+        GUI.set_visible(True, BuildForm.doc)
+        with Lo.ControllerLock():
+            # use a controller lock to lock screen updating.
+            # This will cut down and screen flashiing and add controls faster.
+            BuildForm.doc.addEventListener(self)
+
+            tvc = Write.get_view_cursor(BuildForm.doc)
+            Write.append(tvc, "Building a Form\n")
+            Write.end_paragraph(tvc)
+            
+            self.create_form(BuildForm.doc)
+        Lo.dispatch_cmd("SwitchControlDesignMode")
+        
+        Lo.wait_enter()
+        Lo.close_doc(BuildForm.doc)
+        Lo.close_office()
+
+    def create_form(self, doc: XTextDocument) -> None:
+        props = Forms.add_labelled_control(doc=doc, label="FIRSTNAME", comp_kind=Forms.CompenentKind.TextField, y=11)
```

Comment:
Wondering why this is `doc=doc` and the rest are `doc=BuildForm.doc`? How about define `DOC=BuildForm.doc` then use `doc=DOC`?